Add plotter classes for matplotlib and plotly.

Varunvaruns9 opened the pull request:
<!-- Please fill below the issue number which this code fixes -->
Added basic Matplotlib and Plotly plotter classes with a Plotter class that automatically switches between the two depending on the platform used.

I'll add the animate function in another PR within a few days.

<!-- Please tag any reviewer here -->

pep8speaks commented:
Hello @Varunvaruns9! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`src/einsteinpy/plotting/mplplotter.py`](https://github.com/einsteinpy/einsteinpy/blob/ad2127988ab9d6851a1a5ceaa4beb459d3b462d9/src/einsteinpy/plotting/mplplotter.py):

> [Line 18:121](https://github.com/einsteinpy/einsteinpy/blob/ad2127988ab9d6851a1a5ceaa4beb459d3b462d9/src/einsteinpy/plotting/mplplotter.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 120 characters)

shreyasbapat commented:
Run `tox -e reformat`

Varunvaruns9 commented:
`mypy` is giving an error because I am using `get_ipython` to check if we are using a notebook or not. It gives a not defined error which is what it is supposed to do if run on the command line. I am ignoring that. Is it ok?

shreyasbapat commented on src/einsteinpy/plotting/plotter.py:
```diff
@@ -0,0 +1,70 @@
+from .plotlyplotter import PlotlyPlotter
+
+
+def in_ipynb():
```
Brilliant! 
Yes, you may skip that line :)

shreyasbapat commented on src/einsteinpy/plotting/plotter.py:
```diff
@@ -0,0 +1,70 @@
+        return False
+
+
+class Plotter(MatplotlibPlotter, PlotlyPlotter):
```
Maybe rename it to `GeodesicPlotter` ?

shreyasbapat commented on src/einsteinpy/plotting/plotter.py:
```diff
@@ -0,0 +1,70 @@
+
+    def show(self):
+        if self._notebook:
+            return PlotlyPlotter.show(self)
```
No return here !

shreyasbapat replied:
I would say, do this if-else only once! There's too much redundant code :)

shreyasbapat commented on src/einsteinpy/plotting/plotlyplotter.py:
```diff
@@ -0,0 +1,71 @@
+        r = np.array([coord[1] for coord in vals])
+        phi = np.array([coord[3] for coord in vals])
+
+        x = r * np.cos(phi)
```
This part is wrong! As discussed in #277 
@ritzvik Can you quickly tell how to correct it?

ritzvik replied:
I would say that we should use `return_cartesian=True` when calling `calculate_trajectory`

OR

Use spherical to cartesian conversion using the `coordinates` module for `Schwarzschild`
And BL to cartesian for `Kerr`. To see the implementation just look at any one of the `calculate_trajectory` functions!

shreyasbapat commented on src/einsteinpy/plotting/mplplotter.py:
```diff
@@ -0,0 +1,89 @@
+from matplotlib.patches import Circle
+
+
+class MatplotlibPlotter:
```
Maybe name the class as StaticGeodesicPlotter?  :P

Varunvaruns9 replied:
Should I remove the old ones then?

shreyasbapat replied:
No! Put them in a separate folder. Name that folder, "senile" :P

shreyasbapat replied:
And anyways, put this all in geodesics folder :)

shreyasbapat commented on src/einsteinpy/plotting/plotlyplotter.py:
```diff
@@ -0,0 +1,71 @@
+from plotly.offline import plot as saveplot
+
+
+class PlotlyPlotter:
```
InteractiveGeodesicPlotter ?

Varunvaruns9 commented:
I changed the `GeodesicPlotter` class to work without any extra code.

@ritzvik I added `return_cartesian=True` to `calculate_trajectory()` but it seems to have slowed down creating a new object of class `Geodesic` significantly. Maybe the method can be optimized to work faster?

shreyasbapat commented on src/einsteinpy/plotting/geodesics/__init__.py:
```diff
@@ -1,2 +1,5 @@
-from einsteinpy.plotting.geodesics.geodesics_scatter import ScatterGeodesicPlotter
-from einsteinpy.plotting.geodesics.geodesics_static import StaticGeodesicPlotter
+from einsteinpy.plotting.geodesics.geodesic_plotter import GeodesicPlotter
```
Files name would be 

core.py
static.py
interactive.py

shreyasbapat commented on src/einsteinpy/plotting/__init__.py:
```diff
@@ -1,2 +1,6 @@
+    InteractiveGeodesicPlotter,
+)
+from einsteinpy.plotting.geodesics.staticgeodesic_plotter import StaticGeodesicPlotter
+from einsteinpy.plotting.senile.geodesics_scatter import ScatterGeodesicPlotter
```
This line won't be here

shreyasbapat commented on src/einsteinpy/plotting/geodesics/geodesic_plotter.py:
```diff
@@ -0,0 +1,22 @@
+import random
+
+from .interactivegeodesic_plotter import InteractiveGeodesicPlotter
```
No relative imports

Varunvaruns9 replied:
But if we import a plotting module from another plotting module that raises an error due to circular import.

Varunvaruns9 replied:
But if we try to import a plotting module from another plotting module it would give an error due to circular import. We have to give a relative path in either __init__.py or core.py.

shreyasbapat commented:
Where is the old staticgeodesic? That should also be put in senile

Varunvaruns9 commented:
The old one is almost the same as the new StaticGeodesicPlotter. Do we need to keep that separately?

shreyasbapat commented on docs/source/api/plotting/geodesics/geodesics_scatter.rst:
```diff
@@ -28,5 +28,5 @@ Color
 This module contains the methods for static geodesic plotting using scatter plots.
 
-.. automodule:: einsteinpy.plotting.geodesics.geodesics_scatter
+.. automodule:: einsteinpy.plotting.senile.geodesics_scatter
```
The path to the docs will also be changed!

shreyasbapat replied:
I think you didn't change! The file should lie in 

docs/source/api/plotting/senile/geodesics_scatter.rst

shreyasbapat commented:
> The old one is almost the same as the new StaticGeodesicPlotter. Do we need to keep that separately?

Yes

Varunvaruns9 commented:
@shreyasbapat Did most of the changes except the one with relative paths. Python gives an error due to circular import there.

shreyasbapat commented:
Okay!
